clarify missingness of point estimates; add examples for determining output_type_id format (#197)
hubverse-org/hubDocs: docs/source/user-guide/model-output.md

| EW202242 | weekly rate | 1 | sample | 2 | 3 |
:::

> [^batman]: `NA` (without quotes) indicates missingness in R, which is the expected `output_type_id` for a `mean` `output_type`. 

Author: This might lead people to encode `None` in the CSV itself, which would be interpreted as a character. What about this: 

```suggestion
[^batman]: The `output_type_id` for point estimates (e.g. `mean`) is not applicable. To reflect this, we need to signal that this is a missing value. While a blank cell could indicate missingness, we chose `NA` because it's a clear indicator that it's a presence of an absence and not a mistake. In R, missing values are encoded as `NA` and in Python, they are encoded as `None`.
```

> [^batman]: `NA` (without quotes) indicates missingness in R, which is the expected `output_type_id` for a `mean` `output_type`. 

Reviewer: > This might lead people to encode `None` in the CSV itself, which would be interpreted as a character. What about this:

I take your point! But I find this confusing (perhaps because it's R-centric). Maybe less is more?
```suggestion
[^batman]: The `output_type_id` for point estimates (e.g. `mean`) is not applicable. To
reflect this, we need to signal that this is a missing value. In R, missing values are
encoded as `NA`, and in Python, they are encoded as `None`.
```

> [^batman]: `NA` (without quotes) indicates missingness in R, which is the expected `output_type_id` for a `mean` `output_type`. 

Author: Sold!

>   This is discussed in the [Output type table](#output-type-table)

Reviewer 2: ```suggestion
  This is discussed in the [output type table](#output-type-table)
```

   * Disadvantages:
      * Compatibility: Harder to work with; teams and people who want to work with files need to install additional libraries

Examples of how to create these file formats in R and Python are listed below in

> [the Writing Model Output section](#writing-model-output).

Reviewer 2: ```suggestion
[the writing model output section](#writing-model-output).
```

[^model-id]: When using models for downstream analysis with the [`collect_hub()` function](https://hubverse-org.github.io/hubData/reference/collect_hub.html) in the `hubData` package, one more column called `model_id` is prepended that identifies the model from its filename. 

The model output follows the specification of the `tasks.json` configuration
file of the hub. If you are creating a model and you would like to know what
data type your columns should be in, the hubVerse has utilities to provide and
arrow schema and even a full submission template from the `tasks.json`
configuration file. 

### Arrow Schema

> **The hubverse package `hubData()` has functionality that will generate an

Author: good catch!

| `cdf` | String or numeric: a possible value of the target variable | Numeric between 0.0 and 1.0: the value of the cumulative distribution function of the predictive distribution at the value of the outcome variable specified by the output_type_id |
| `pmf` | String naming a possible category of a discrete outcome variable | Numeric between 0.0 and 1.0: the value of the probability mass function of the predictive distribution when evaluated at a specified level of a categorical outcome variable. |
| `sample` | Positive integer sample index | Numeric: a sample from the predictive distribution.
:::

:::{note} 
:name: output-type-caveats

The model output type IDs have different caveats depending on the `output_type`:

`mean` and `median`
: Point estimates do not have an `output_type_id` because you can only have one
point estimate for each combination of task IDs. However, because the
`output_type_id` column is required, something has to go in this place, which
is a missing value. This is encoded as [`NA` in

> R](https://www.njtierney.com/post/2020/09/17/missing-flavour/) (which is why
> our schemas prior to 4.0.0 encoded these as `["NA"]`) and `None` in Python. See

Reviewer: ```suggestion
R](https://www.njtierney.com/post/2020/09/17/missing-flavour/) and `None` in Python. See
```
I think this reads better without the aside about schema versions...the important thing is to convey what the `output_type_id` column will look like for point estimate tasks

Author: totes agree

of writing model output to a hub in R and Python using parquet and CSV files.
In these examples, we are assuming the following variables already exist:

 - `model_out` is the tabular output from your model formatted as specified
   in [the formats of model output section](#formats-of-model-output).
 - `hub_path` is the path to the hub cloned on your local computer
 - `model_id` is the combination of `<team_abbr>-<model_abbr>` 
 - `file_name` is the file name of your model formatted as
   `<round_id>-<model_id>.csv` (or parquet)

(example-csv)=
### Example: model output as CSV

To write to CSV, you would use the `write_csv()` from the `readr` package in R and
the `to_csv()` method in Python. 

#### Writing CSV with R

```r
# ... generate model data ...
outfile <- fs::path(hub_path, "model-output", model_id, file_name)
readr::write_csv(model_out, outfile)
```

#### Writing CSV with Python

```python
import pandas as pd
import os.path
# ... generate model data ...
outfile = os.path.join(hub_path, "model-output", "team1-modelA", model_id)
model_out.to_csv(outfile, index = False, na_rep = "NA")
```

(example-parquet)=
### Example: model output as parquet

> Writing to parquet is similar as writing to CSV, but with the caveat that you
> additionally need to ensure that the `output_type_id` column matches the
> [expected `output_type_id_datatype` property of the schema](#output-type-id-datatype).

Reviewer: ```suggestion
Unlike a CSV, a parquet files contain embedded information about the data types of its
columns. Therefore, when writing model output files as parquet, it's critical that you
first ensure the data type of the `output_type_id` column matches the
[expected `output_type_id_datatype` property of the schema](#output-type-id-datatype).

If the data types of the model output parquet file don't match the hub's schema, the
submission will not validate.
```

Tried to surface the reason for the difference between csv and parquet, without getting too in-the-weeds. Also think it might be helpful to clarify why it's important to mind the parquet column types (maybe that would be a good message for one of those fancy warning boxes?)

Author: I think this is much better wording!

In practice, you will need to know whether or not the expected data type is a
**string/character**, **float/numeric**, or an **Int/integer**.

#### Writing parquet with R

```r
# ... generate model data ...
outfile <- fs::path(hub_path, "model-output", model_id, file_name)

> model_out$output_type_id <- as.character(model_out$output_type_id) # or as.numeric(), or as.integer()

Reviewer: ```suggestion
# update the output_type_id data type to match the hub's schema
model_out$output_type_id <- as.character(model_out$output_type_id) # or as.numeric(), or as.integer()
```

arrow::write_parquet(model_out, outfile)
```

#### Writing parquet with Python

```python
import pandas as pd
import os.path
# ... generate model data ...
outfile = os.path.join(hub_path, "model-output", "team1-modelA", model_id)

> model_out["output_type_id"] = model_out["output_type_id"].astype("string") # or "float", or "Int64"

Reviewer: ```suggestion
# update the output_type_id data type to match the hub's schema
model_out["output_type_id"] = model_out["output_type_id"].astype("string") # or "float", or "Int64"
```

_Model outputs are a specially formatted tabular representation of predictions._
Each row corresponds to a unique prediction, and each column provides information about what is being predicted, its scope, and its value. 
Per hubverse convention, **there are two groups of columns providing metadata about the prediction**[^model-id], followed by **a value column with the actual output**. Each group of columns serves a specific purpose: (1) the **"task ID"** columns provide details about what is being predicted, and (2) the two **"model output representation"** columns specify the type of prediction and identifying information about that prediction. Finally, (3) the **value** column provides the model output of the prediction. [Details about the column specifications](column-details) can be found below.

> [^model-id]: When using models for downstream analysis with the [`collect_hub()` function](https://hubverse-org.github.io/hubData/reference/collect_hub.html) in the `hubData` package, one more column called `model_id` is prepended added that identifies the model from its filename. 

Reviewer 3: ```suggestion
[^model-id]: When using models for downstream analysis with the [`collect_hub()` function](https://hubverse-org.github.io/hubData/reference/collect_hub.html) in the `hubData` package, one more column called `model_id` is prepended that identifies the model from its filename. 
```

> The model output follows the specification of the `tasks.json` configuration

Reviewer 3: While the following section is useful and detailed, it feels like it should be more of an aside towards the end of the discussion (i.e. it flows after the section where we tell folks that - btw your missing values need to be typed!). It is also inevitably very R centric when in some ways we're trying to be a little more cross-platform in this section. 

I think it would actually be better to have a way hubs can include schema information in a cross-platform way in their documentation, like a function that prints a table to their README. As I mentioned though this is a separate issue and more generally related to, in my opinion, work we need to do make metadata associated with hub data more visible/accessible.

So for now, I think we should keep this section as it's got lots of useful info but maybe move it down and link it to the mention of typed missing values.

Author: Yeah, I wasn't super happy with the placement of this and I agree with your sentiment. This is further evidence I need to split the documentation into the Diátaxis framework. 

> I think it would actually be better to have a way hubs can include schema information in a cross-platform way in their documentation, like a function that prints a table to their README. As I mentioned though this is a separate issue and more generally related to, in my opinion, work we need to do make metadata associated with hub data more visible/accessible.

Y'know, I wouldn't be surprised if this could be achieved with some vanilla in-browser JS.

Reviewer 3: Oooh that could be cool! Maybe we should open an issue somewhere to track this? (Not sure where the best place would be...)

Author: I opened it in the site builder dashboard repository.

file of the hub. If you are creating a model and you would like to know what
data type your columns should be in, the hubVerse has utilities to provide and
arrow schema and even a full submission template from the `tasks.json`
configuration file. 

### Arrow Schema

**The hubverse package `hubData()` has functionality that will generate an
arrow schema so that you can ensure your output matches the expected type.** 

Here is some example code that can help. In this example, `tasks_path` is the
path to your `tasks.json` config file (note: it can also be a URL like [the
simple forecasts example hub `tasks.json`
file](https://raw.githubusercontent.com/hubverse-org/example-simple-forecast-hub/refs/heads/main/hub-config/tasks.json))

```r
# read the configuration file and get the latest round

> config_tasks <- hubData::read_config_file(tasks_path)

Reviewer 3: You could use `hubData::read_config(hub_path)` instead and just provide the path to the hub

Author: I was thinking about @trobacker's comment that he usually models outside of the context of a hub, but I can certainly change this.

Reviewer 3: They still need access to the hub to get to the config file though? The only difference between the two functions is that one takes the full path to the config file and the other the path to a hub and then builds the config path relative to the root of the hub. It also defaults to the `tasks.json` which is the config most commonly loaded.

> config_tasks <- hubData::read_config_file(tasks_path)

Author: 🙈

schema <- hubData::create_hub_schema(config_tasks)
```

The schema output will look something like this:

```
Schema
origin_date: date32[day]
target: string
horizon: int32
location: string
age_group: string
output_type: string
output_type_id: double
value: int32
model_id: string
```

### Submission Template

**The hubverse package `hubValidations()` has functionality
that will generate template data to get your started.** This submission
template can be written as a CSV or parquet file and then imported in to
whatever software you use to run your model.

Here is some example code that can help. In this example, `tasks_path` is the
path to your `tasks.json` config file (note: it can also be a URL like [the
simple forecasts example hub `tasks.json`
file](https://raw.githubusercontent.com/hubverse-org/example-simple-forecast-hub/refs/heads/main/hub-config/tasks.json))

```r
# read the configuration file and get the latest round
config_tasks <- hubData::read_config_file(tasks_path)
rounds <- hubData::get_round_ids(config_tasks)
this_round <- rounds[length(rounds)]
# create the submission template (this may take some time if your submission uses samples)

> tmpl <- hubValidations::submission_tmpl(config_tasks, round_id = this_round)

Author: Thank you! The assertion on `hub_con` was actually a failure from my earlier code where I did not name the argument.

R](https://www.njtierney.com/post/2020/09/17/missing-flavour/) (which is why
our schemas prior to 4.0.0 encoded these as `["NA"]`) and `None` in Python. See
[The example on writing parquet files](#example-parquet) for details.

`pmf`
: Values are required to sum to 1 across all
`output_type_id` values within each combination of values of task ID variables.
This representation should only be used if the outcome variable is truly
discrete; a CDF representation is preferred if the categories represent a
binned discretization of an underlying continuous variable.

`sample`
: Depending on the hub specification, samples with the same sample index
(specified by the `output_type_id`) may be assumed to correspond to a single
sample from a joint distribution across multiple levels of the task ID
variables — further details are discussed below.

`cdf` (and `pmf` for ordinal variables)
: In the hub's `tasks.json` configuration file, the values of the
`output_type_id` should be listed in order from low to high.

:::

## 

(writing-model-output)=
## Writing model output to a hub

When submitting model output to a hub, it should be placed in a folder with the

> name of your model in the model outputs folder specified by the hub

Reviewer 3: ```suggestion
name of your `model_id` in the model outputs folder specified by the hub
```
`model_id` is not the same as model name. It's the combination of `team_abbr` and `model_abbr`. This is discussed at the top of the page and might be a good idea to link to it.

Author: Oh good point! I did not realize this!

administrator (this is usually called `model-output`). Below are two examples
of writing model output to a hub in R and Python using parquet and CSV files.
In these examples, we are assuming the following variables already exist:

 - `model_out` is the tabular output from your model formatted as specified
   in [the formats of model output section](#formats-of-model-output).

>  - `hub_path` is the path to the hub cloned on your local computer

Reviewer 3: ```suggestion
 - `hub_path` is the path to the hub cloned on your local computer
 - `model_id` is the combination of `<team_abbr>-<model_abbr>` 
```

 - `file_name` is the file name of your model formatted as
   `<round_id>-<model_id>.csv` (or parquet)

(example-csv)=
### Example: model output as CSV

To write to CSV, you would use the `write_csv()` from the `readr` package in R and
the `to_csv()` method in Python. 

#### Writing CSV with R

```r

> library("fs")

Reviewer 3: ```suggestion
```

> library("readr")

Reviewer 3: ```suggestion
```
For conciseness

# ... generate model data ...

> outfile <- path(hub_path, "model-output", "team1-modelA", model_id)

Reviewer 3: ```suggestion
outfile <- fs::path(hub_path, "model-output", model_id, file_name)
```

> write_csv(model_out, outfile)

Reviewer 3: ```suggestion
readr::write_csv(model_out, outfile)
```

```

#### Writing CSV with Python

```python
import pandas as pd
import os.path
# ... generate model data ...
outfile = os.path.join(hub_path, "model-output", "team1-modelA", model_id)

> model_out.to_csv(outfile, index = False, na_rep = "NA")

Author: Yeah it's nuanced and annoying. I should probably just leave it out and allow NA's to be blanks. R should be able to handle them.

> model_out.to_csv(outfile, index = False, na_rep = "NA")

Author: I've looked at this again and I've removed the `na_rep` argument to simplify the example and it works.

```

(example-parquet)=
### Example: model output as parquet

Writing to parquet is similar as writing to CSV, but with the caveat that you
additionally need to ensure that the `output_type_id` column matches the
[expected `output_type_id_datatype` property of the schema](#output-type-id-datatype).
In practice, you will need to know whether or not the expected data type is a
**string/character**, **float/numeric**, or an **Int/integer**.

#### Writing parquet with R

```r

> library("fs")

Reviewer 3: ```suggestion
```

> library("arrow")

Reviewer 3: ```suggestion
```

# ... generate model data ...

> outfile <- path(hub_path, "model-output", "team1-modelA", model_id)

Reviewer 3: ```suggestion
outfile <- fs::path(hub_path, "model-output", model_id, file_name)
```

The model output follows the specification of the `tasks.json` configuration
file of the hub. If you are creating a model and would like to know what
data type your columns should be in, the Hubverse has utilities to provide [an
arrow schema](#arrow-schema) and even a [full submission
template](#submission-template) from the `tasks.json` configuration file. 

When submitting model output to a hub, it should be placed in a folder with the
name of your `model_id` in the model outputs folder specified by the hub
administrator (this is usually called `model-output`). Below are R and Python
examples for writing Hubverse-compliant model output files in both CSV and
parquet format. In these examples, we are assuming the following variables
already exist:

 - `hub_path` is the path to the hub cloned on your local computer
 - `model_id` is the combination of `<team_abbr>-<model_abbr>`
 - `file_name` is the file name of your model formatted as
   `<round_id>-<model_id>.csv` (or `.parquet`)
 - `model_out` is the tabular output from your model formatted as specified
   in [the formats of model output section](#formats-of-model-output).

(submission-template)=
### Submission Template

> **The hubverse package `hubValidations()` has functionality

Reviewer 3: ```suggestion
**The hubverse package `hubValidations` has functionality
```
As its a pkg name and not a function name, we don't need parentheses

Reviewer 4: I nearly suggested it but was not sure I was correct, only function need parentheses, no? (Just to be sure I have the correct standard)

#### Writing CSV with R

When writing a model output file in R, use the `readr` package.
```r
# ... generate model data ...
outfile <- fs::path(hub_path, "model-output", model_id, file_name)
readr::write_csv(model_out, outfile)
```

#### Writing CSV with Python

Python users can use the `pandas` package when creating CSV model output files.
```python
import pandas as pd
import os.path

# ... generate model data ...
outfile = os.path.join(hub_path, "model-output", model_id, file_name)
model_out.to_csv(outfile, index = False)
```

(example-parquet)=
### Example: model output as parquet

Unlike a CSV, a parquet files contain embedded information about the data types
of its columns.
Therefore, when writing model output files as parquet, it's
critical that you first ensure the data type of your columns matches the
expected type from [the Arrow schema](#arrow-schema).

If the data types of the model output parquet file don't match the hub's schema, the
submission will not validate.
In practice, you will need to know whether or not the expected data type is a
**string/character**, **float/numeric**, or an **Int/integer**.

(arrow-schema)=
#### Arrow Schema

> **The hubverse packages `hubData()` and `hubUtils()` have functionality that will generate an

Reviewer 3: ```suggestion
**The hubverse packages `hubData` and `hubUtils` have functionality that will generate an
```

arrow schema so that you can ensure your output matches the expected type.** 

Here is some example code that can help. In this example, `hub_path` is the
path to the hub on your local machine.

```r
# read the configuration file and get the latest round
config_tasks <- hubUtils::read_config(hub_path, "tasks")
schema <- hubData::create_hub_schema(config_tasks)
```

The schema output will look something like this:

```
Schema
origin_date: date32[day]
target: string
horizon: int32
location: string
age_group: string
output_type: string
output_type_id: double
value: int32
model_id: string
```

> 

Reviewer 3: ```suggestion

If you have a data frame of model output data (e.g. `model_out`) and want to coerce the column data types to the hub schema, you can also use [`hubData::coerce_to_hub_schema()`](https://hubverse-org.github.io/hubData/reference/coerce_to_hub_schema.html), e.g. 
```r
hubData::coerce_to_hub_schema(model_out, config_tasks)
```

```

#### Writing parquet with R

```r
# ... generate model data ...
outfile <- fs::path(hub_path, "model-output", model_id, file_name)

# update the output_type_id data type to match the hub's schema

> model_out$output_type_id <- as.character(model_out$output_type_id) # or as.numeric(), or as.integer()

Author: oh good point!

arrow::write_parquet(model_out, outfile)
```

#### Writing parquet with Python

```python
import pandas as pd
import os.path
# ... generate model data ...
outfile = os.path.join(hub_path, "model-output", model_id, file_name)

# update the output_type_id data type to match the hub's schema

> model_out["output_type_id"] = model_out["output_type_id"].astype("string") # or "float", or "Int64"

Author: thank you!

`output_type_id` values within each combination of values of task ID variables.
This representation should only be used if the outcome variable is truly
discrete; a CDF representation is preferred if the categories represent a
binned discretization of an underlying continuous variable.

`sample`
: Depending on the hub specification, samples with the same sample index
(specified by the `output_type_id`) may be assumed to correspond to a single
sample from a joint distribution across multiple levels of the task ID
variables — further details are discussed below.

`cdf` (and `pmf` for ordinal variables)
: In the hub's `tasks.json` configuration file, the values of the
`output_type_id` should be listed in order from low to high.

:::

(writing-model-output)=
## Writing model output to a hub

When submitting model output to a hub, it should be placed in a folder with the
name of your model in the model outputs folder specified by the hub
administrator (this is usually called `model-output`). Below are two examples
of writing model output to a hub in R and Python using parquet and CSV files.
In these examples, we are assuming the following variables already exist:

>  - `model_out_tbl` is the tabular output from your model formatted as specified

Reviewer 3: This is not quite accurate. Model submissions do not generally contain `model_id` columns, When date is accessed from a hub via `coonect_hub()` it does. Only data accessed by a hub is a `model_out_tbl` class object.

Author: I'll go ahead and change it. I think when I wrote this, I did not fully realize that `model_out_tbl` was a reserved name.

   in [the formats of model output section](#formats-of-model-output).

>  - `path_to_hub` is the path to the hub cloned on your local computer

Reviewer 3: ```suggestion
 - `hub_path` is the path to the hub cloned on your local computer
```
This is the argument name across the hubverse

>  - `model_name` is the file name of your model formatted as

Reviewer 3: ```suggestion
 - `file_name` is the file name of your model formatted as
```

>    `<round_id>-<model_name>.csv` (or parquet)

Reviewer 3: ```suggestion
   `<round_id>-<model_id>.csv` (or parquet)
```
For consistency with hubverse terminology

(example-csv)=
### Example: model output as CSV

To write to CSV, you would use the `write_csv()` from the `readr` package in R and
the `to_csv()` method in Python. 

#### Writing CSV with R

> ```r
> library("fs")
> library("readr")
> # ... generate model data ...
> outfile <- path(path_to_hub, "model-output", "team1-modelA", model_name)
> write_csv(model_out_tbl, outfile)
> ```
> 
> #### Writing CSV with Python
> 
> ```python
> import pandas as pd
> import os.path
> # ... generate model data ...
> outfile = os.path.join(path_to_hub, "model-output", "team1-modelA", model_name)
> model_out_tbl.to_csv(outfile, index = False, na_rep = "NA")
> ```

Author: ````suggestion
```r
library("fs")
library("readr")
# ... generate model data ...
outfile <- path(hub_path, "model-output", "team1-modelA", model_id)
write_csv(model_out_tbl, outfile)
```

#### Writing CSV with Python

```python
import pandas as pd
import os.path
# ... generate model data ...
outfile = os.path.join(hub_path, "model-output", "team1-modelA", model_id)
model_out_tbl.to_csv(outfile, index = False, na_rep = "NA")
```
````

(example-parquet)=
### Example: model output as parquet

Writing to parquet is similar as writing to CSV, but with the caveat that you
additionally need to ensure that the `output_type_id` column matches the

> [expected `output_type_id_datatype` property of the schema](#output-type-id-datatype).

Reviewer 3: That's only if the hub has specified an `output_type_id_datatype`. It might be just auto detected. See https://hubverse.io/en/latest/user-guide/model-output.html#the-output-type-id-column-data-type

This also might need to be performed regardless of the format being written out?. E.g is a column written out from python that contains `NA`s or `None` plus integers still written out as integers?

Author: > E.g is a column written out from python that contains NAs or None plus integers still read into R as integers?

Oh good catch! I had updated https://github.com/hubverse-org/hubDocs/issues/198#issuecomment-2430287818 with my findings, but I forgot to update here (in which pandas needs to convert to `Int64` to allow for integer `None` values). 

**The answer to your question is yes**. Even if a parquet file is written with python (either polars or pandas), the type stays with the parquet file and a `null` value inside an integer column is represented as an integer `NA` in R. 

The only thing that seems to have a problem with integer `null` values is pandas. By default, if it reads in a parquet file that has `null` values in an integer column, that column will be coerced to a float (unless the `dtype_backend` is set to `pyarrow`).

Reviewer 3: I think we should make it clear that `NA`s might also pop up in task IDs, not just the `output_type_id` column, for example [when a certain task ID is relevant to one modeling task but not another](https://hubverse.io/en/latest/quickstart-hub-admin/tasks-config.html#required-and-optional-elements) (also discussed in https://github.com/hubverse-org/hubValidations/issues/140)

In practice, you will need to know whether or not the expected data type is a

> **string/character** or a **float/numeric**.

Reviewer 3: Related to this PR but not expected to be resolved by it, it's clear submitting teams need to have an idea of the hub schema, especally of the `output_type_id` column datatype. While `hubData::create_hub_schema()` and even `hubValidations::submission_tbl()` can provide this info in R, I'm wondering if there's a better way to communicate this infor to submitting teams more broadly?

Author: I think this is exactly the information I needed to know! I was initially confused by the use of `hubData::create_hub_schema()` since I thought it was related to our schemas repository. I had not used these functions and I've learned that some modelers were unaware that hubValidations even existed. At the moment, I think most people are scanning the schema to figure out how to format their data, but the table would likely be helpful and I think it's worthwhile to add into this section. 

As an aside, I wonder if we could add in a function somewhere (maybe in `hubData`) that just provides the expected `output_type_id` datatype given the schema like:

```r
output_type_id_datatype()
#> "numeric"
```

Author: I went ahead and iterated on this, rearranging some things and creating examples.

#### Writing parquet with R

> 
> ```r
> library("fs")
> library("arrow")
> # ... generate model data ...
> outfile <- path(path_to_hub, "model-output", "team1-modelA", model_name)
> model_out_tbl$output_type_id <- as.character(model_out_tbl$output_type_id) # or as.numeric()
> arrow::write_parquet(model_out_tbl, outfile)
> ```
> 
> 
> #### Writing parquet with Python
> 
> ```python
> import pandas as pd
> import os.path
> # ... generate model data ...
> outfile = os.path.join(path_to_hub, "model-output", "team1-modelA", model_name)
> model_out_tbl["output_type_id"] = model_out_tbl["output_type_id"].astype("string") # or "float"
> model_out_tbl.to_parquet(outfile)
> ```

Author: ````suggestion

```r
library("fs")
library("arrow")
# ... generate model data ...
outfile <- path(hub_path, "model-output", "team1-modelA", model_id)
model_out_tbl$output_type_id <- as.character(model_out_tbl$output_type_id) # or as.numeric()
arrow::write_parquet(model_out_tbl, outfile)
```

#### Writing parquet with Python

```python
import pandas as pd
import os.path
# ... generate model data ...
outfile = os.path.join(hub_path, "model-output", "team1-modelA", model_id)
model_out_tbl["output_type_id"] = model_out_tbl["output_type_id"].astype("string") # or "float"
model_out_tbl.to_parquet(outfile)
```
````

| `output_type` | `output_type_id` | `value` |
| ------ | ------ | ------ | 
| `mean` | `NA`/`None` (not used for mean predictions) | Numeric: the mean of the predictive distribution |
| `median` | `NA`/`None` (not used for median predictions) | Numeric: the median of the predictive distribution |
| `quantile` | Numeric between 0.0 and 1.0: a probability level | Numeric: the quantile of the predictive distribution at the probability level specified by the output_type_id |
| `cdf` | String or numeric: a possible value of the target variable | Numeric between 0.0 and 1.0: the value of the cumulative distribution function of the predictive distribution at the value of the outcome variable specified by the output_type_id |
| `pmf` | String naming a possible category of a discrete outcome variable | Numeric between 0.0 and 1.0: the value of the probability mass function of the predictive distribution when evaluated at a specified level of a categorical outcome variable. |
| `sample` | Positive integer sample index | Numeric: a sample from the predictive distribution.
:::

:::{note} 
:name: output-type-caveats

> The model output type ids have different caveats depending on the `output_type`:

Reviewer 2: ```suggestion
The model output type IDs have different caveats depending on the `output_type`:
```

This representation should only be used if the outcome variable is truly
discrete; a CDF representation is preferred if the categories represent a
binned discretization of an underlying continuous variable.

`sample`
: Depending on the hub specification, samples with the same sample index
(specified by the `output_type_id`) may be assumed to correspond to a single
sample from a joint distribution across multiple levels of the task ID
variables — further details are discussed below.

`cdf` (and `pmf` for ordinal variables)
: In the hub's `tasks.json` configuration file, the values of the
`output_type_id` should be listed in order from low to high.

:::

(writing-model-output)=
## Writing model output to a hub

When submitting model output to a hub, it should be placed in a folder with the
name of your model in the model outputs folder specified by the hub
administrator (this is usually called `model-output`). Below are two examples
of writing model output to a hub in R and Python using parquet and CSV files.
In these examples, we are assuming the following variables already exist:

 - `model_out_tbl` is the tabular output from your model formatted as specified

>    in [the Formats of Model Output section](#formats-of-model-output).

Reviewer 2: ```suggestion
   in [the formats of model output section](#formats-of-model-output).
```

 - `path_to_hub` is the path to the hub cloned on your local computer
 - `model_name` is the file name of your model formatted as
   `<round_id>-<model_name>.csv` (or parquet)

(example-csv)=
### Example: model output as CSV

> To write to CSV, you would use the `write_csv()` from the readr package in R and

Reviewer 2: ```suggestion
To write to CSV, you would use the `write_csv()` from the `readr` package in R and
```

:::{table} Relationship between the three model output representation columns with respect to the type of prediction (`output_type`)
| `output_type` | `output_type_id` | `value` |
| ------ | ------ | ------ | 

> | `mean` | None[^missingno] (not used for mean predictions) | Numeric: the mean of the predictive distribution |

Reviewer 3: I would prefer at the very least something like `NA/None` (without any quotation marks) here with a prompt to more language specific details. That way `NA` is still easily recognisable as a correct value by R users. I believe `<NA>` also exists in python via `pd.NA`, is the closest thing to R's `NA` and I _think_ that's actually what @MKupperman used in his successful exploration (but I might be wrong).

In any case, I've renamed and moved #198 to here and have suggested we test out the various missingness options available in python so we have clarity about what works and what doesn't when we write this part of the documentation. Details in the issue.

> | `median` | None (not used for median predictions) | Numeric: the median of the predictive distribution |

Reviewer 3: Same comment as above

| `quantile` | Numeric between 0.0 and 1.0: a probability level | Numeric: the quantile of the predictive distribution at the probability level specified by the output_type_id |
| `cdf`[^cdf] | String or numeric: a possible value of the target variable | Numeric between 0.0 and 1.0: the value of the cumulative distribution function of the predictive distribution at the value of the outcome variable specified by the output_type_id |
| `pmf`[^pmf] | String naming a possible category of a discrete outcome variable | Numeric between 0.0 and 1.0: the value of the probability mass function of the predictive distribution when evaluated at a specified level of a categorical outcome variable.[^cdf] |
| `sample`[^sample] | Positive integer sample index | Numeric: a sample from the predictive distribution.
:::

[^missingno]: Point estimates don't have an `output_type_id` because you can only have one point estimate for each combination of task IDs.
   However, because the `output_type_id` column is requrired, something has to go in this place, which is a missing value. 

>    This is encoded as [`NA_character_` in R](https://www.njtierney.com/post/2020/09/17/missing-flavour/) (which is why our schemas prior to 4.0.0 encoded these as `["NA"]`).

Reviewer 3: This is not always encoded as `NA_character_`. Output type ID columns may well be `double` if e.g. only `mean` and `quantile` output type IDs are being collected. The data type will be [determined by the overall config](https://hubverse.io/en/latest/user-guide/model-output.html#the-output-type-id-column-data-type). What we want to make clear is that `output_type_id` columns containing `NA` values must still adhere to the hub schema. Then we can provide tips on how to do this is different languages. E.g. in R you can assign `NA_real_` or  `NA_character_` or coerce to any other data type using the `as.*()` functions.

Also the `"NA"` in the config doesn't really have anything to do with a data type.  It is actually read in as the default logical:
``` r
jsonlite::fromJSON('{"required": ["NA"]}')$required |> typeof()
#> [1] "logical"
```
It's just the way to trick R to read an `NA` in as a bare `NA` is not allowed in JSON

We'd need to do e.g.:
``` r
jsonlite::fromJSON('{"required": ["NA_character_"]}')$required |> typeof()
#> [1] "character"
```
to get character but that's even more R specific so we avoided it.

Reviewer 3: Overall I feel this section is important enough and needs enough detailed discussion that it's own section is appropriate rather than a footnote

Author: > Overall I feel this section is important enough and needs enough detailed discussion that it's own section is appropriate rather than a footnote

Yeah I was realizing that nearly every cell in this table had a footnote, which meant that a new section is absolutely needed.

>    **If you use Python** to write parquet files, the `output_type_id` column should be an array of `None` and you will need to [explicitly cast the `output_type_id` column as a "string"](https://github.com/hubverse-org/hubValidations/issues/131#issuecomment-2427654006) (via Pandas):    

Reviewer 3: Not necessarily the entire column but any output type ID values associated with a point estimate output type. This may be the entire column if only point estimate output types are being submitted.

Reviewer 3: Also I don't feel a link to an issue comment is the best way to discuss this. As discussed above, I think a full section with more detailed advice would be more appropriate
